internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +100,210 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
-	return errorsutil.NewAggregate(errs)
+	return errorsutil.NewAggregate(applicationErrors)
 }
 
-func (c *gitHubWatcherController) validateUpstream(ctx controllerlib.Context, upstream *v1alpha1.GitHubIdentityProvider) (*upstreamgithub.ProviderConfig, error) {
-	result := upstreamgithub.ProviderConfig{
-		Name: upstream.Name,
+func (c *gitHubWatcherController) validateUpstreamAndUpdateConditions(ctx controllerlib.Context, upstream *v1alpha1.GitHubIdentityProvider) (
+	config *upstreamgithub.ProviderConfig, // If validated, returns the config
+	applicationErr error) { // This error will only refer to programmatic errors such as inability to perform a Dial or dereference a pointer, not configuration errors
+
+	conditions := make([]*metav1.Condition, 0)
+	applicationErrors := make([]error, 0)
+	var clientID, clientSecret string
+	secret, applicationErr := c.secretInformer.Lister().Secrets(upstream.Namespace).Get(upstream.Spec.Client.SecretName)
+	if applicationErr != nil {
+		applicationErrors = append(applicationErrors, applicationErr)
+	}
+	if applicationErr != nil ||
+		secret == nil ||
+		secret.Type != gitHubClientSecretType ||
+		len(secret.Data[clientIDDataKey]) < 1 ||
+		len(secret.Data[clientSecretDataKey]) < 1 {
+		conditions = append(conditions, &metav1.Condition{
+			Type:   ClientCredentialsObtained,
+			Status: metav1.ConditionFalse,
+			Reason: upstreamwatchers.ReasonNotFound,
+			Message: fmt.Sprintf("secret with name %q must be found in namespace %q with type %q and keys %q and %q",
+				upstream.Spec.Client.SecretName,
+				upstream.Namespace,
+				gitHubClientSecretType,
+				clientIDDataKey,
+				clientSecretDataKey),
+		})
+	} else if applicationErr == nil {
+		clientID = string(secret.Data[clientIDDataKey])
+		clientSecret = string(secret.Data[clientSecretDataKey])
+		conditions = append(conditions, &metav1.Condition{
+			Type:    ClientCredentialsObtained,
+			Status:  metav1.ConditionTrue,
+			Reason:  upstreamwatchers.ReasonSuccess,
+			Message: "client credentials have been obtained, although they may not yet have been validated",
+		})
+	}
+
+	// Should not happen due to CRD defaulting
+	host := ""
+	if upstream.Spec.GitHubAPI.Host == nil || len(*upstream.Spec.GitHubAPI.Host) < 1 {
+		conditions = append(conditions, &metav1.Condition{
+			Type:    HostValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  "InvalidHost",
+			Message: `spec.githubAPI.host ("") is not valid`,
+		})
+		conditions = append(conditions, &metav1.Condition{
+			Type:    GitHubConnectionValid,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "UnableToValidate",
+			Message: `unable to determine if spec.githubAPI.host ("") is reachable or if TLS verification succeeds`,
+		})
+		conditions = append(conditions, &metav1.Condition{
+			Type:    TLSConfigurationValid,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "UnableToValidate",
+			Message: `tls configuration cannot be validated for spec.githubAPI.host ("")`,
```

Comment:
For `Status: metav1.ConditionUnknown,` we've been using this [message consistently](https://github.com/vmware-tanzu/pinniped/blob/main/internal/controller/authenticator/jwtcachefiller/jwtcachefiller.go#L72) in other controllers:

```
msgUnableToValidate = "unable to validate; see other conditions for details"
```

to say "The REAL problem you have to fix is in the condition that has a root cause problem".

Reply:
done